### PR TITLE
[WIP] Add utilities to build and manage recursive representations

### PR DESCRIPTION
**Pull request by rrodenbusch**
Added
-    qiskit/utils/reprbuild.py
-    qiskit/utils/reprparse.py

Implement a QuantumCircuit representation recursivley including:
-    CircuitInstruction, Instruction
-    ControlledGate, Gate,
-    Qubit, QuantumRegister,
-    Clbit, ClassicalRegister,
-    Register

Added example code, examples/python/build_repr.py, to rebuild and compare objects, including:
-    Bell Circuit
-    cx gate
-    cx instruction

### Summary
Address Issue [8594 ](https://github.com/Qiskit/qiskit-terra/issues/8594) with formatting and manipulation utilities for better \_\_repr\_\_ generation.  

Implement tools to build recursive representations that provide a single, unambiguous, authoritative representation of any variable within a system.  The tools attempt to balance the python standards regarding the machine readable (__repr__) representation, human readable (__str__) representation, and the recreation of equivalent objects via eval( repr( _obj_ ) ).  Tools to build, parse, print and recreate an equivalent instance from the recursive representations are provided.  When using the utilities:
- repr(_obj_) results in a string providing safer evaluation and less complexity than fully recreating an object
- eval(repr(_obj_)) results in a python list containing a summary and object details for attributes included
- if rebuild methods are implemented (see example code), the ReprParser class can reconstruct equivalent objects
- recursive representation do not require modification to __repr__ and __str__ for included attributes (see Qubit and QuantumRegister in example)
- additional information can be added to the summary to make the machine readable representation more user human friendly (see QuantumCircuit)

### Details and comments
See attached markdown file and zipped jupyter notebook (notebook and html output) for examples. 

The following issues are relevant to this discussion:
- Qiskit-terra: Add good \_\_repr\_\_ methods to all public classes [#8594 ](https://github.com/Qiskit/qiskit-terra/issues/8594)
- Qiskit-terra: add QuantumCircuit.from_instructions [#9006](https://github.com/Qiskit/qiskit-terra/pull/9006) and [#8709](https://github.com/Qiskit/qiskit-terra/issues/8709)
- QAMP: Add good \_\_repr\_\_ methods to all public classes [#37](https://github.com/qiskit-advocate/qamp-fall-22/issues/37)

**Included:**

1. Implementations of \_\_repr\_\_ for QuantumCircuit, ControlledGate, Gate, Clbit, and ClassicalRegister.  The QuantumCircuit representation customizes the summary section of the machine readable representation to be more user friendly.  **_['class: QuantumCircuit,name: Inner, 2 qubits, 2 clbits, 4 instructions', {....._**
2. Definition of attribute lists for QuantumCircuit, CircuitInstruction, Instruction, ControlledGate, Gate, Qubit, QuantumRegister, Clbit, ClassicalRegister, and Register
3. Implementations of  \_\_str\_\_ method for CircuitInstruction, Instruction, ControlledGate, Gate, Clbit, ClassicalRegister
4. Leaves untouched pre-existing methods for QuantumCircuit.\_\_str\_\_, CircuitInstruction.\_\_repr\_\_, Instruction.\_\_repr\_\_, Register.\_\_str\_\_

Note, currently the following tests will fail if \_\_str\_\_ is configured in Qubit, QuantumRegister or if \_\_repr\_\_ is configured in Qubit, QuantumRegister, Register:
- test.python.transpiler.test_layout.LayoutTest.test_layout_repr_with_holes
- test.python.transpiler.test_layout.LayoutTest.test_layout_repr,  

WIP Tasks Remaining:
- Address test case failure for Qubit,QuantumRegister and Register in test_layout_repr_with_holes
- Address test case failure for Qubit,QuantumRegister and Register in test_layout_repr
- Build test cases, using examples as starting point
- Release Notes

Attachments:

[utils-build-repr-jupyter-Notebook.zip](https://github.com/Qiskit/qiskit-terra/files/9882669/utils-build-repr-jupyter-Notebook.zip)
[README.md](https://github.com/Qiskit/qiskit-terra/files/9882850/README.md)

### CODE REVIEW

**Review by jakelishman — CHANGES_REQUESTED** [review]
Thank you very much for the interest in contributing!

This is certainly an impressive amount of work, but I'm sorry, it's not really what we're looking for to solve the issues you've linked to.  I explained in those issues what the outputs should look like (approximately), and other commenters also mentioned why attempting to make all `repr`'s re-parseable can often cause them to fail in their main purpose, which is to give the programmer a short, legible overview of what the object is.  This PR really goes a very different direction, and one that's much too complex for our need.

I'm sorry, but the approach is just fundamentally not what we wanted, and I don't see any path towards merging this.  I do appreciate that it must have been a lot of work, and I don't want to discourage you from contributing in the future.  Please note that it's _really_ a lot better to discuss direction with us on "feature request" issues before starting off on so much work like this, to avoid such a mismatch happening again.

**Review by 1ucian0 — CHANGES_REQUESTED** [review]
I second the @jakelishman review. This seems like an overkill for a simple issue and it comes at a cost. A representation should be simple as it is used for debugging and developing.The scaffolding around something like this has the risk of complicate the debugging and error understanding.

A good example of that is that this code might raise an exception in `__repr__` time, which might be called already inside another exception.

I would even argue that recurrent `__repr__` calling might be a bit too much. If `A` has instances of `B`, you might not want to call `B.__repr__` from `A.__repr__`, since `B.__repr__` might give too much detail in the context of `A`.

So, bottom line, just defining  `__repr__` ad-hoc in each class does the job and "simple is better than complex."

**Comment by rrodenbusch** [comment]
Totally understandable and thanks for the quick feedback. This is an upgrade to a library I have used in jupyter for doing large dataset analysis, so I ran with it.  The process to get here has been quite educational regarding the codebase and the dev-test environment, plus my library is more refined now; so I don't feel like it's been wasted effort on my part.  I will turn these into the simpler reprs and submit them that way.

Any feedback on the form, structure or content of the PR itself?

**Comment by 1ucian0** [comment]
> I will turn these into the simpler reprs and submit them that way.

Thanks! In that case I will close this PR and open a fresh one. You can link back to this one if for context, is needed.

> any feedback on the form, structure or content of the PR itself?

Hard to know as a WIP. Assuming this PR would be ready for review, some general comments:
 *  it does too much on possible unrelated fronts. For example, the `__str__` changes seems unrelated with `__repr__`. Smaller PRs means easier reviews.
 * No test of the added code? Maybe some of the examples can be repurpose for this. However, I have the impression they can be simpler (we prefer things that test single functionalities, when possible) 
 * No release note?
 * Good job at documenting added functions/methods.
